File: src/extras-attica/libmetadata.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_Attica__Metadata
#include <QByteArray>
#include <QList>
#include <QPair>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <metadata.h>
#include "libmetadata.h"
#include "libmetadata.hxx"

Attica__Metadata* Attica__Metadata_new() {
    return new Attica::Metadata();
}

Attica__Metadata* Attica__Metadata_new2(const Attica__Metadata* other) {
    return new Attica::Metadata(*other);
}

void Attica__Metadata_OperatorAssign(Attica__Metadata* self, const Attica__Metadata* other) {
    self->operator=(*other);
}

int Attica__Metadata_Error(const Attica__Metadata* self) {
    return static_cast<int>(self->error());
}

void Attica__Metadata_SetError(Attica__Metadata* self, int errorVal) {
    self->setError(static_cast<Attica::Metadata::Error>(errorVal));
}

int Attica__Metadata_StatusCode(const Attica__Metadata* self) {
    return self->statusCode();
}

void Attica__Metadata_SetStatusCode(Attica__Metadata* self, int code) {
    self->setStatusCode(static_cast<int>(code));
}

libqt_string Attica__Metadata_StatusString(const Attica__Metadata* self) {
    QString _ret = self->statusString();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Metadata_SetStatusString(Attica__Metadata* self, const libqt_string status) {
    QString status_QString = QString::fromUtf8(status.data, status.len);
    self->setStatusString(status_QString);
}

libqt_string Attica__Metadata_Message(Attica__Metadata* self) {
    QString _ret = self->message();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Metadata_SetMessage(Attica__Metadata* self, const libqt_string message) {
    QString message_QString = QString::fromUtf8(message.data, message.len);
    self->setMessage(message_QString);
}

int Attica__Metadata_TotalItems(Attica__Metadata* self) {
    return self->totalItems();
}

void Attica__Metadata_SetTotalItems(Attica__Metadata* self, int items) {
    self->setTotalItems(static_cast<int>(items));
}

int Attica__Metadata_ItemsPerPage(Attica__Metadata* self) {
    return self->itemsPerPage();
}

void Attica__Metadata_SetItemsPerPage(Attica__Metadata* self, int itemsPerPage) {
    self->setItemsPerPage(static_cast<int>(itemsPerPage));
}

libqt_string Attica__Metadata_ResultingId(Attica__Metadata* self) {
    QString _ret = self->resultingId();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void Attica__Metadata_SetResultingId(Attica__Metadata* self, const libqt_string id) {
    QString id_QString = QString::fromUtf8(id.data, id.len);
    self->setResultingId(id_QString);
}

libqt_list /* of libqt_pair  tuple of libqt_string and libqt_string  */ Attica__Metadata_Headers(const Attica__Metadata* self) {
    QList<QPair<QByteArray, QByteArray>> _ret = self->headers();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_pair /* tuple of libqt_string and libqt_string */* _arr = static_cast<libqt_pair /* tuple of libqt_string and libqt_string */*>(malloc(sizeof(libqt_pair /* tuple of libqt_string and libqt_string */) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QPair<QByteArray, QByteArray> _lv_ret = _ret[i];
        // Convert QPair<> from C++ memory to manually-managed C memory
        libqt_string* _lv_first = static_cast<libqt_string*>(malloc(sizeof(libqt_string)));
        libqt_string* _lv_second = static_cast<libqt_string*>(malloc(sizeof(libqt_string)));
        QByteArray _lv_first_qb = _lv_ret.first;
        libqt_string _lv_first_str;
        _lv_first_str.len = _lv_first_qb.length();
        _lv_first_str.data = static_cast<const char*>(malloc(_lv_first_str.len + 1));
        memcpy((void*)_lv_first_str.data, _lv_first_qb.data(), _lv_first_str.len);
        ((char*)_lv_first_str.data)[_lv_first_str.len] = '\0';
        *_lv_first = _lv_first_str;
        QByteArray _lv_second_qb = _lv_ret.second;
        libqt_string _lv_second_str;
        _lv_second_str.len = _lv_second_qb.length();
        _lv_second_str.data = static_cast<const char*>(malloc(_lv_second_str.len + 1));
        memcpy((void*)_lv_second_str.data, _lv_second_qb.data(), _lv_second_str.len);
        ((char*)_lv_second_str.data)[_lv_second_str.len] = '\0';
        *_lv_second = _lv_second_str;
        libqt_pair _lv_out;
        _lv_out.first = static_cast<void*>(_lv_first);
        _lv_out.second = static_cast<void*>(_lv_second);
        _arr[i] = _lv_out;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void Attica__Metadata_Delete(Attica__Metadata* self) {
    delete self;
}

File: src/extras-attica/libmetadata.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const metadata_enums = enums;
const std = @import("std");
pub const struct_u8_u8 = extern struct { first: []u8, second: []u8 };

/// https://api.kde.org/attica-metadata.html
pub const attica__metadata = struct {
    /// New constructs a new Attica::Metadata object.
    ///
    ///
    pub fn New() QtC.Attica__Metadata {
        return qtc.Attica__Metadata_new();
    }

    /// New2 constructs a new Attica::Metadata object.
    ///
    /// ``` other: QtC.Attica__Metadata ```
    pub fn New2(other: ?*anyopaque) QtC.Attica__Metadata {
        return qtc.Attica__Metadata_new2(@ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#operator-eq)
    ///
    /// ``` self: QtC.Attica__Metadata, other: QtC.Attica__Metadata ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.Attica__Metadata_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#error)
    ///
    /// ``` self: QtC.Attica__Metadata ```
    ///
    /// Returns: ``` metadata_enums.Error ```
    pub fn Error(self: ?*anyopaque) i32 {
        return qtc.Attica__Metadata_Error(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setError)
    ///
    /// ``` self: QtC.Attica__Metadata, errorVal: metadata_enums.Error ```
    pub fn SetError(self: ?*anyopaque, errorVal: i32) void {
        qtc.Attica__Metadata_SetError(@ptrCast(self), @intCast(errorVal));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#statusCode)
    ///
    /// ``` self: QtC.Attica__Metadata ```
    pub fn StatusCode(self: ?*anyopaque) i32 {
        return qtc.Attica__Metadata_StatusCode(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setStatusCode)
    ///
    /// ``` self: QtC.Attica__Metadata, code: i32 ```
    pub fn SetStatusCode(self: ?*anyopaque, code: i32) void {
        qtc.Attica__Metadata_SetStatusCode(@ptrCast(self), @intCast(code));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#statusString)
    ///
    /// ``` self: QtC.Attica__Metadata, allocator: std.mem.Allocator ```
    pub fn StatusString(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Metadata_StatusString(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::metadata.StatusString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setStatusString)
    ///
    /// ``` self: QtC.Attica__Metadata, status: []const u8 ```
    pub fn SetStatusString(self: ?*anyopaque, status: []const u8) void {
        const status_str = qtc.libqt_string{
            .len = status.len,
            .data = status.ptr,
        };
        qtc.Attica__Metadata_SetStatusString(@ptrCast(self), status_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#message)
    ///
    /// ``` self: QtC.Attica__Metadata, allocator: std.mem.Allocator ```
    pub fn Message(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Metadata_Message(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::metadata.Message: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setMessage)
    ///
    /// ``` self: QtC.Attica__Metadata, message: []const u8 ```
    pub fn SetMessage(self: ?*anyopaque, message: []const u8) void {
        const message_str = qtc.libqt_string{
            .len = message.len,
            .data = message.ptr,
        };
        qtc.Attica__Metadata_SetMessage(@ptrCast(self), message_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#totalItems)
    ///
    /// ``` self: QtC.Attica__Metadata ```
    pub fn TotalItems(self: ?*anyopaque) i32 {
        return qtc.Attica__Metadata_TotalItems(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setTotalItems)
    ///
    /// ``` self: QtC.Attica__Metadata, items: i32 ```
    pub fn SetTotalItems(self: ?*anyopaque, items: i32) void {
        qtc.Attica__Metadata_SetTotalItems(@ptrCast(self), @intCast(items));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#itemsPerPage)
    ///
    /// ``` self: QtC.Attica__Metadata ```
    pub fn ItemsPerPage(self: ?*anyopaque) i32 {
        return qtc.Attica__Metadata_ItemsPerPage(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setItemsPerPage)
    ///
    /// ``` self: QtC.Attica__Metadata, itemsPerPage: i32 ```
    pub fn SetItemsPerPage(self: ?*anyopaque, itemsPerPage: i32) void {
        qtc.Attica__Metadata_SetItemsPerPage(@ptrCast(self), @intCast(itemsPerPage));
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#resultingId)
    ///
    /// ``` self: QtC.Attica__Metadata, allocator: std.mem.Allocator ```
    pub fn ResultingId(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.Attica__Metadata_ResultingId(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("attica::metadata.ResultingId: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#setResultingId)
    ///
    /// ``` self: QtC.Attica__Metadata, id: []const u8 ```
    pub fn SetResultingId(self: ?*anyopaque, id: []const u8) void {
        const id_str = qtc.libqt_string{
            .len = id.len,
            .data = id.ptr,
        };
        qtc.Attica__Metadata_SetResultingId(@ptrCast(self), id_str);
    }

    /// [Qt documentation](https://api.kde.org/attica-metadata.html#headers)
    ///
    /// ``` self: QtC.Attica__Metadata, allocator: std.mem.Allocator ```
    pub fn Headers(self: ?*anyopaque, allocator: std.mem.Allocator) []struct_u8_u8 {
        const _arr: qtc.libqt_list = qtc.Attica__Metadata_Headers(@ptrCast(self));
        defer {
            const _pair: [*]qtc.libqt_pair = @ptrCast(@alignCast(_arr.data));
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_pair[i].first));
                qtc.libqt_free(_pair[i].first);

                qtc.libqt_string_free(@ptrCast(&_pair[i].second));
                qtc.libqt_free(_pair[i].second);
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc(struct_u8_u8, _arr.len) catch @panic("attica::metadata.Headers: Memory allocation failed");
        const _data: [*]struct_u8_u8 = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.Attica__Metadata ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.Attica__Metadata_Delete(@ptrCast(self));
    }
};

/// https://api.kde.org/attica-metadata.html#types
pub const enums = struct {
    pub const Error = enum {
        pub const NoError: i32 = 0;
        pub const NetworkError: i32 = 1;
        pub const OcsError: i32 = 2;
    };
};
